### ISSUE TITLE
page_config.json no longer encapsulated in python env

### ISSUE THREAD

**Issue by ajbozarth** [jupyterlab/jupyterlab#9240]
## Description

In Lab 2.X page_config.json is keep within the python env path, allowing for a user to run lab in multiple python env in parallel.

In Lab 3.0 page_config.json is now stored in the user home, and is shared by all python envs, which can cause issue when working with multiple envs with different expected configs.

**Comment by ajbozarth** [comment]
@blink1073

**Comment by jasongrout** [comment]
I thought now the page config used the Jupyter directory hierarchy, so it *could* be at the environment level?

**Comment by jasongrout** [comment]
So...put it where you want? If you want it at the env level, put it there. Otherwise put it in the user level.

Or are you saying that page_config.json gets automatically put somewhere by a process out of your control?

**Comment by ajbozarth** [comment]
exactly, it gets automatically moved to the first location on the list, which is usually the user dir

**Comment by jasongrout** [comment]
> it gets automatically moved to the first location on the list

Sorry, I'm missing context here. What moves it automatically? How are you creating this page_config.json?

**Comment by ajbozarth** [comment]
@blink1073 can explain it better, I'm just reiterating what he told me, but iiuc we moved from just creating and using the page_config.json directly to using it to instantiate and persist the configs that are handled by traitlets now. The traitlets are handled through jupyter_core libraries, which by default use the first config dir listed by `jupyter --path` to store the persisted page_config.json, which is usually the user dir. 

In short our code moves the page_config.json automatically, but it chooses the dir to move it to based on the jupyter_core libraries leveraged by using traitlets.

**Comment by jasongrout** [comment]
Just to be clear, can you give instructions that will create a page_config.json in my user config directory automatically without me manually creating it?

**Comment by jasongrout** [comment]
Also, I spent some time looking into how to solve this user/env/system problem once and for all. For example, we now can set the `JUPYTER_PATH` and `JUPYTER_CONFIG_PATH` environment variables, which take precedence over the user directories. I think that was added so environments could insert themselves in before the user directory. I tried to see how to make conda always set these variables appropriately any time it entered any environment. After poking around in the conda code for a bit, I gave up.

I think it would be much easier to have an environment variable flag, like `JUPYTER_SYS_FIRST` (with a better name :), that switches the user and sys-prefix levels once and for all.

**Comment by ajbozarth** [comment]
you just have to disable an extension using the cli tool, eg:
`jupyter labextension disable @jupyterlab/launcher-extension`
and a `page_config.json` will be created in your user dir:
`ls ~/.jupyter/labconfig/page_config.json`

**Comment by ajbozarth** [comment]
> For example, we now can set the JUPYTER_PATH and JUPYTER_CONFIG_PATH environment variables

This is a way to "fix" this, but it seemed "hacky" to @blink1073 and me, thus why he was going to look for a better solution

**Comment by jasongrout** [comment]
I just filed https://github.com/jupyter/jupyter_core/pull/199, which I've been meaning to write for years now...

**Comment by ajbozarth** [comment]
Based on that PR would we have lab set that env var or would we just document it and expect users to set it?

**Comment by jasongrout** [comment]
I think users would set it for whatever situation they were in. Users not working in environments could naturally expect page_config to be user-level, for example.

**Comment by jasongrout** [comment]
Regardless of jupyter/jupyter_core#199, we should still figure out what to do for lab 3.0, since we can't assume that everyone has a new version of jupyter_core installed, even if we can get a new release there before lab 3.0.

What options do we have? For example, we could explicitly put that page config in the environment directory for now?

**Comment by ajbozarth** [comment]
is the jupyter_core dependency version not set by jupyterlab?

**Comment by jasongrout** [comment]
> is the jupyter_core dependency version not set by jupyterlab?

One question is if we can get a jupyter_core version out before jlab 3.0, I think.

Huh, apparently we also don't list jupyter_core as a dependency. See #9250 9251
